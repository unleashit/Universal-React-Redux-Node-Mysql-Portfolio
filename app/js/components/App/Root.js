import React, { Component } from 'react';

class Root extends Component {
    renderInitialState() {
        if (this.props.initialState) {
            const innerHtml = `window.__INITIAL_STATE__ = ${JSON.stringify(
                this.props.initialState
            )}`;
            return <script dangerouslySetInnerHTML={{ __html: innerHtml }} />;
        }
    }

    renderEnvironment() {
        const innerHtml = `window.__ENVIRONMENT__ = '${__ENVIRONMENT__};'
            window.__GOOGLE_ANALYTICS__ = '${__GOOGLE_ANALYTICS__}';
            window.__LIVE_CHAT_ADMIN_NAME__ = '${__LIVE_CHAT_ADMIN_NAME__}';`;
        return <script dangerouslySetInnerHTML={{ __html: innerHtml }} />;
    }

    lazyLoadScripts() {
        const innerHtml = `
                function lazyLoadScript(scrpt) {
                    var element = document.createElement("script");
                    element.src = scrpt;
                    element.async = true;
                    document.body.appendChild(element);
                }`;
        return <script dangerouslySetInnerHTML={{ __html: innerHtml }} />;
    }

    lazyLoadScript(scrpt) {
        if (process.env.NODE_ENV !== 'production' && scrpt.global) {
            return <script src={scrpt.devSrc} />;
        } else {
            const innerHtml = `
                window.addEventListener('load', lazyLoadScript.bind(null, '${
                    scrpt.src
                }'), false);`;
            return <script dangerouslySetInnerHTML={{ __html: innerHtml }} />;
        }
    }

    render() {
        const head = this.props.head;
        const attrs = head.htmlAttributes.toComponent();

        return (
            <html {...attrs}>
                <head>
                    {head.title.toComponent()}
                    {head.meta.toComponent()}
                    <link
                        rel="apple-touch-icon"
                        sizes="180x180"
                        href="/images/favicons/apple-touch-icon.png"
                    />
                    <link
                        rel="icon"
                        type="image/png"
                        href="/images/favicons/favicon-32x32.png"
                        sizes="32x32"
                    />
                    <link
                        rel="icon"
                        type="image/png"
                        href="/images/favicons/android-chrome-192x192.png"
                        sizes="192x192"
                    />
                    <link
                        rel="icon"
                        type="image/png"
                        href="/images/favicons/favicon-16x16.png"
                        sizes="16x16"
                    />
                    <link
                        rel="manifest"
                        href="/images/favicons/manifest.json"
                    />
                    <link
                        rel="mask-icon"
                        href="/images/favicons/safari-pinned-tab.svg"
                        color="#5bbad5"
                    />
                    <link
                        rel="shortcut icon"
                        href="/images/favicons/favicon.ico"
                    />
                    <meta
                        name="google-site-verification"
                        content="PcZbHamu8N77ConcEm3oykmUEroNsJhpM6ZTrnvEO40"
                    />
                    <meta name="format-detection" content="telephone=no" />
                    <meta
                        name="msapplication-config"
                        content="/images/favicons/browserconfig.xml"
                    />
                    <meta name="theme-color" content="#ffffff" />
                    <meta name="referrer" content="origin" />
                    {head.link.toComponent()}
                    {this.renderEnvironment()}
                    {this.lazyLoadScripts()}
                </head>
                <body>
                    <div
                        id="root"
                        dangerouslySetInnerHTML={{ __html: this.props.content }}
                    />
                    {this.renderInitialState()}
                    {head.script.toComponent()}
                    {this.lazyLoadScript({
                        src: '/js/global.min.js',
                        devSrc: '/js/global.js',
                        global: true
                    })}
                    {this.lazyLoadScript({
                        src:
                            'https://cdnjs.cloudflare.com/ajax/libs/smooth-scroll/11.0.2/js/smooth-scroll.min.js'
                    })}
                </body>
            </html>
        );
    }
}

export default Root;
