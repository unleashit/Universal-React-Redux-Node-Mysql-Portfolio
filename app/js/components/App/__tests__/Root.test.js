import Root from '../Root';

describe('<Root />', () => {
    let wrapper;
    const props = {
        head: {
            htmlAttributes: {
                toComponent: jest.fn(),
            },
            title: {
                toComponent: jest.fn(),
            },
            meta: {
                toComponent: jest.fn(),
            },
            link: {
                toComponent: jest.fn(),
            },
            script: {
                toComponent: jest.fn(),
            },
        },
        content: '',
        initialState: null,
    };

    // make it output the production JS
    process.env.NODE_ENV = 'production';

    // global vars are set by server in app.js so stub them
    global.__ENVIRONMENT__ = 'test';
    global.__LIVE_CHAT_ADMIN_NAME__ = 'test';

    beforeEach(() => {
        wrapper = shallow(<Root {...props} />);
    });

    afterAll(() => {
        process.env.NODE_ENV = 'test';
    });

    test('Renders parent HTML', () => {
        expect(wrapper.find('html')).toHaveLength(1);
    });

    test('Matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('renderInitialState()', () => {
        props.initialState = { test: 'test' };
        wrapper = shallow(<Root {...props} />);
        const results = wrapper.instance().renderInitialState();
        expect(results.props.dangerouslySetInnerHTML).toEqual({
            __html: 'window.__INITIAL_STATE__ = {"test":"test"}',
        });
    });

    test('lazyLoadScript() returns script with right attrs', () => {
        process.env.NODE_ENV = 'development';
        let results;

        results = wrapper.instance().lazyLoadScript({
            src: '/js/global.min.js',
            devSrc: '/js/global.js',
            global: true,
        });
        expect(results.props.src).toEqual('/js/global.js');

        process.env.NODE_ENV = 'production';
        results = wrapper.instance().lazyLoadScript({
            src: '/js/global.min.js',
            devSrc: '/js/global.js',
            global: true,
        });
        expect(results.props.dangerouslySetInnerHTML.__html).toContain(
            "window.addEventListener('load', lazyLoadScript.bind(null, '/js/global.min.js'), false);"
        );
    });
});
