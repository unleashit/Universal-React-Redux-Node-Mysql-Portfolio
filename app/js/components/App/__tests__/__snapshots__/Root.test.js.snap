// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Root /> Matches snapshot 1`] = `
<html>
  <head>
    <link
      href="/images/favicons/apple-touch-icon.png"
      rel="apple-touch-icon"
      sizes="180x180"
    />
    <link
      href="/images/favicons/favicon-32x32.png"
      rel="icon"
      sizes="32x32"
      type="image/png"
    />
    <link
      href="/images/favicons/android-chrome-192x192.png"
      rel="icon"
      sizes="192x192"
      type="image/png"
    />
    <link
      href="/images/favicons/favicon-16x16.png"
      rel="icon"
      sizes="16x16"
      type="image/png"
    />
    <link
      href="/images/favicons/manifest.json"
      rel="manifest"
    />
    <link
      color="#5bbad5"
      href="/images/favicons/safari-pinned-tab.svg"
      rel="mask-icon"
    />
    <link
      href="/images/favicons/favicon.ico"
      rel="shortcut icon"
    />
    <meta
      content="PcZbHamu8N77ConcEm3oykmUEroNsJhpM6ZTrnvEO40"
      name="google-site-verification"
    />
    <meta
      content="telephone=no"
      name="format-detection"
    />
    <meta
      content="/images/favicons/browserconfig.xml"
      name="msapplication-config"
    />
    <meta
      content="#ffffff"
      name="theme-color"
    />
    <meta
      content="origin"
      name="referrer"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "window.__ENVIRONMENT__ = 'test;'
            window.__LIVE_CHAT_ADMIN_NAME__ = 'test';",
        }
      }
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "
                function lazyLoadScript(scrpt) {
                    var element = document.createElement(\\"script\\");
                    element.src = scrpt;
                    element.async = true;
                    document.body.appendChild(element);
                }",
        }
      }
    />
  </head>
  <body>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
      id="root"
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "
                window.addEventListener('load', lazyLoadScript.bind(null, '/js/global.min.js'), false);",
        }
      }
    />
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "
                window.addEventListener('load', lazyLoadScript.bind(null, 'https://cdnjs.cloudflare.com/ajax/libs/smooth-scroll/11.0.2/js/smooth-scroll.min.js'), false);",
        }
      }
    />
  </body>
</html>
`;
